### DISCUSSION TITLE
Model of the two problems?

### DISCUSSION THREAD

**Discussion by fsikora** [ds4dm/ml4co-competition#18] [Q&A] [answered]
Hi all,
is it on purpose that the linear model of the two known problem is not given? 
Best,
Florian

**Comment by gasse** [comment]
Hi Florian,

We only provide the problem instances in the form of MPS files, which you can inspect to get an understanding of the structure of each problem.

If you want to use the linear models to make decisions within your agent, you can extract the linear model of each problem within your observation function. For example, in the primal and dual tasks, you can use the following methods from the PySCIPOpt API to fetch the local LP rows and columns and reconstruct the local linear problem:
```
m = model.as_pyscipopt()
cols = [var.getCol() for var in m.getVars(transformed=True) if var.isInLP()]  # to get local LP columns
rows = m.getLPRowsData()  # to get local LP rows

# for each row in the LP (lhs <= wTx + cst <= rhs)
row.getLhs()       # left-hand-side of the row (lhs)
row.getRhs()       # right-hand-side of the row (rhs)
row.getConstant()  # constant of a row (cst)
row.getCols()      # columns with a non-zero entry in the row (i where w_i != 0)
row.getVals()      # the corresponding non-zero entries (w_i where w_i != 0)

# for each column in the LP (x_i)
col.getLPPos()         # position (id) of the column in the LP
col.getLb()            # lower bound of the column (lb <= x_i)
col.getUb()            # upper bound of the column (x_i <= ub)
col.getVar().getObj()  # objective coefficient of the column (c_i)
```

In the config task, because the LP has not been constructed yet when your agent is called, you can instead use the [MilpBipartite](https://doc.ecole.ai/py/en/stable/reference/observations.html#milp-bipartite) observation function that ships with in Ecole to get the original LP formulation (before preprocessing). Note that unfortunately this observation function can not be used in the `primal` or `dual` tasks at the moment.

For help on how to build your own observation function, and how to re-use an existing one, please refer to Ecole's [documentation](https://doc.ecole.ai/py/en/stable/howto/create-functions.html).

Hope that helps, let us know if you need more help.

Best,
Maxime

**Reply by fsikora** [reply]
Dear Maxime,

thanks for the answer!
I don't understand, you don't want or you can't give the general model? :) In case of the former, why only one problem is anonymous? Are we supposed to infer the model from the informal english description given?

Best,
Florian

**Reply by gasse** [reply]
Hi @fsikora ,

I understand that you would like a description of the instances, not just the linear model.

I realize we haven't really put a lot of effort into the description of the internal structure of the instances themselves. I don't see why we couldn't do that, at least for the two first benchmarks. The explanation is more that this aspect has been put aside a little bit during the preparation phase really. I'll see what we can do to provide a short description of the variables / constraints and their meaning, so that you don't have to decipher each instance by hand. This should come in a few weeks maybe.

Best,
Maxime

**Reply by harshadkhadilkar** [reply]
Hi Maxime, all the "row.*" commands in your suggestion work, but the row.getCols() returns a set of objects such as <pyscipopt.scip.Column at 0x7f07b0c880f0>. Any hint on how to get the column ID from this?

**Reply by gasse** [reply]
Those are PySCIPOpt objects of the `column` class, whose API you can find [here](https://github.com/scipopt/PySCIPOpt/blob/20580884e39907b00cd8b7f5a55c6356e6b30622/src/pyscipopt/scip.pyx#L317). You can then access column identifiers as follows:
```python
col.getLPPos()  # position of column in current LP, or -1 if it is not in LP
col.getVar().getIndex()   # unique index of the variable in the SCIP problem
```

Best,
Maxime

**Reply by harshadkhadilkar** [reply]
Thank you!

**Comment by gasse** [comment, marked as answer]
Hi @fsikora ,

We've finally put together a more precise description of the two first benchmarks, that is, how each problem instance is modeled as a MILP. You'll find the description here: https://github.com/ds4dm/ml4co-competition/blob/main/DATA.md

Hope that helps,

Best,
Maxime

**Reply by fsikora** [reply]
thanks a lot !
